Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1843 @@
+  constructor
+  intro c f h
+  by_cases hc : c = 0
+  · exact Or.inl hc
+  · right
+    ext x
+    rw [LocallyConstant.ext_iff] at h
+    apply_fun fun (y : ℤ) ↦ c * y
+    · simp at h
+      simp
+      exact h x
+    · exact mul_right_injective₀ hc
+
+theorem GoodProducts.linearIndependentSingleton :
+    LinearIndependent ℤ (eval ({fun _ ↦ false} : Set (I → Bool))) := by
+  refine linearIndependent_unique (eval ({fun _ ↦ false} : Set (I → Bool))) ?_
+  simp only [eval, Products.eval, List.map, List.prod_nil, ne_eq, one_ne_zero, not_false_eq_true]
+
+end Zero
+
+section Maps
+
+theorem contained_eq_proj (o : Ordinal) (h : contained C o) :
+    C = C.proj (ord I · < o) := by
+  have := proj_prop_eq_self C (ord I · < o)
+  simp [Set.proj, Bool.not_eq_false] at this
+  exact (this (fun i x hx ↦ h x hx i)).symm
+
+theorem isClosed_proj (o : Ordinal) (hC : IsClosed C) : IsClosed (C.proj (ord I · < o)) :=
+  (continuous_proj (ord I · < o)).isClosedMap C hC
+
+theorem contained_proj (o : Ordinal) : contained (C.proj (ord I · < o)) o := by
+  intro x hx j hj
+  obtain ⟨_, ⟨_, h⟩⟩ := hx
+  dsimp [Proj] at h
+  rw [← congr_fun h j] at hj
+  simp only [Bool.ite_eq_true_distrib, if_false_right_eq_and] at hj
+  exact hj.1
+
+/-- The `ℤ`-linear map induced by precomposition of the projection `C → C.proj (ord I · < o)`. -/
+@[simps!]
+noncomputable
+def πs (o : Ordinal) : LocallyConstant (C.proj (ord I · < o)) ℤ →ₗ[ℤ] LocallyConstant C ℤ :=
+  LocallyConstant.comapₗ ℤ (ProjRestrict C (ord I · < o)) (continuous_projRestrict _ _)
+
+theorem coe_πs (o : Ordinal) (f : LocallyConstant (C.proj (ord I · < o)) ℤ) :
+    πs C o f = f ∘ ProjRestrict C (ord I · < o) := by
+  simp only [πs_apply, continuous_projRestrict, LocallyConstant.coe_comap]
+
+theorem injective_πs (o : Ordinal) : Function.Injective (πs C o) :=
+  LocallyConstant.comap_injective _ (continuous_projRestrict _ _) (surjective_projRestrict _ _)
+
+/-- The `ℤ`-linear map induced by precomposition of the projection
+    `C.proj (ord I · < o₂) → C.proj (ord I · < o₁)` for `o₁ ≤ o₂`. -/
+@[simps!]
+noncomputable
+def πs' {o₁ o₂ : Ordinal} (h : o₁ ≤ o₂) :
+    LocallyConstant (C.proj (ord I · < o₁)) ℤ →ₗ[ℤ] LocallyConstant (C.proj (ord I · < o₂)) ℤ :=
+  LocallyConstant.comapₗ ℤ (ProjRestricts C (fun _ hh ↦ lt_of_lt_of_le hh h))
+    (continuous_projRestricts _ _)
+
+theorem coe_πs' {o₁ o₂ : Ordinal} (h : o₁ ≤ o₂) (f : LocallyConstant (C.proj (ord I · < o₁)) ℤ) :
+    (πs' C h f).toFun = f.toFun ∘ (ProjRestricts C (fun _ hh ↦ lt_of_lt_of_le hh h)) := by
+  simp only [πs'_apply, LocallyConstant.toFun_eq_coe, continuous_projRestricts,
+    LocallyConstant.coe_comap]
+
+theorem injective_πs' {o₁ o₂ : Ordinal} (h : o₁ ≤ o₂) : Function.Injective (πs' C h) :=
+  LocallyConstant.comap_injective _ (continuous_projRestricts _ _) (surjective_projRestricts _ _)
+
+end Maps
+
+section ProductsFactorisation
+
+namespace Products
+
+theorem lt_of_head!_lt {l : Products I} {o : Ordinal}
```

Comment:
This looks like another place where some extra list/lex/ordinal API can make this proof a 1-liner